source/Components/List.cpp

```diff
@@ -0,0 +1,131 @@
+/*
+** Project FileShare, 2022
+**
+** Author Léo Lhuile
+**
+** Started on  Sun Feb 13 22:03:32 2022 Léo Lhuile
+** Last update Mon Aug 29 20:45:51 2022 Léo Lhuile
+**
+** List.cpp : Implementation of List class
+*/
+
+#include <algorithm>
+#include "Components/List.hpp"
+
+namespace FileShare::GUI::Components {
+    List::List(const char* typeName, bool initRenderer)
+        : tgui::Panel(typeName, initRenderer)
+    {
+    }
+
+    List::~List() {}
+
+    void List::setSpaceBetweenItems(unsigned int space)
+    {
+        this->space = space;
+        this->updateItemsPosition();
+    }
+
+    void List::setAutoHeight(bool enable)
+    {
+        this->autoHeight = enable;
+        this->updateHeight();
+    }
+
+    void List::add(const tgui::Widget::Ptr &item, const tgui::String &widgetName)
+    {
+        tgui::Panel::add(item, widgetName);
+
+        auto widgetIndex = this->getWidgets().size() - 1;
+
+        item->setAutoLayout(tgui::AutoLayout::Manual);
+        item->setAutoLayoutUpdateEnabled(false);
+        item->setWidth("100%");
+        this->onSizeChangeSignals.push_back(
+            item->onSizeChange.connect([this, widgetIndex]() {
+                this->updateItemsPosition(widgetIndex);
+                this->updateHeight();
+            })
+        );
+        this->updateItemsPosition(widgetIndex);
+        this->updateHeight();
+    }
+
+    bool List::remove(const Widget::Ptr &widget)
+    {
+        auto widgets = this->getWidgets();
+        auto widgetIndex = std::find(widgets.begin(), widgets.end(), widget) - widgets.begin();
+
+        auto success = tgui::Panel::remove(widget);
+
+        if (success) {
+            widget->onSizeChange.disconnect(this->onSizeChangeSignals[widgetIndex]);
+            this->onSizeChangeSignals.erase(this->onSizeChangeSignals.begin() + widgetIndex);
+            this->updateItemsPosition(widgetIndex - 1);
+            this->updateHeight();
+        }
+
+        return success;
+    }
+
+    void List::addSpacer(unsigned int height)
+    {
+        auto spacer = tgui::Panel::create();
+        spacer->setHeight(height);
+        spacer->getRenderer()->setBackgroundColor(tgui::Color::Transparent);
+
+        this->add(spacer);
+    }
+
+    void List::addSeparator(const tgui::Layout2d &size, tgui::Color color)
+    {
+        auto separator = tgui::SeparatorLine::create();
+
+        this->add(separator);
+
+        separator->setSize(size);
+        separator->setPosition("parent.width / 2 - width / 2", separator->getPosition().y);
+        separator->getRenderer()->setColor(color);
+    }
+
+    void List::updateItemsPosition(unsigned int from)
+    {
+        auto items = this->getWidgets();
+        if (items.size() == 0) {
```

Comment:
```suggestion
        if (from >= items.size()) {
```
Avoid crashes if misused. Also handle the `== 0` case, since if size == 0, then from cannot be smaller